### PR TITLE
introduce rcutils_strcasecmp, case insensitive string compare.

### PR DESCRIPTION
**Pull request by fujitatomoya**
close https://github.com/ros2/rcutils/issues/279

Signed-off-by: Tomoya.Fujita <Tomoya.Fujita@sony.com>

### CODE REVIEW

**Comment by fujitatomoya** [comment]
@iuhilnehc-ynos @Ada-King
CC: @clalancette 

either of you, could you take a look at this? it seems that https://github.com/ros2/rcl/pull/741 does not require this anymore, but i guess that this would be useful.

File: CMakeLists.txt [hunk 1]
```diff
@@ -64,6 +64,7 @@ set(rcutils_sources
   src/repl_str.c
   src/shared_library.c
   src/snprintf.c
+  src/strcasecmp.c
```
Comment (iuhilnehc-ynos) [inline CMakeLists.txt:67]: What about moving it after `src/split.c`?

File: include/rcutils/strcasecmp.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+#ifndef RCUTILS__STRCASECMP_H_
+#define RCUTILS__STRCASECMP_H_
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include <string.h>
```
Comment (iuhilnehc-ynos) [inline include/rcutils/strcasecmp.h:23]: Is it good to put this line here? I'd like to move it to '.c'

File: src/strcasecmp.c [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
```
Comment (iuhilnehc-ynos) [inline src/strcasecmp.c:18]: I think we don't expect c++ compiler to compile it. ( I know some files also added these lines, but others did not. )

File: include/rcutils/strcasecmp.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ * \param[in] s2 String with null terminated to compare.
+ * \return An integer less than, equal to, or greater than zero if s1 is, after
+ *   ignoring case, found to be less than, to match, or be greater than s2,
+ *   respectively.
+ */
+RCUTILS_PUBLIC
+RCUTILS_WARN_UNUSED
+int
+rcutils_strcasecmp(const char * s1, const char * s2);
+
```
Comment (iuhilnehc-ynos) [inline include/rcutils/strcasecmp.h:44]: I think it's better to add `strncasecmp(const char *s1, const char * s2, size_t n)` more.

File: test/test_strcasecmp.cpp [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include "rcutils/strcasecmp.h"
+
+// Tests the rcutils_strcasecmp() function.
+TEST(TestStrcasecmp, test_strcasecmp) {
+  EXPECT_EQ(0, rcutils_strcasecmp("", ""));
```
Comment (iuhilnehc-ynos) [inline test/test_strcasecmp.cpp:21]: Do we need to add some test cases if arguments are NULL.

**Review by iuhilnehc-ynos — COMMENTED** [review]
@fujitatomoya 

nitpick, just some minor comments.

File: src/strcasecmp.c [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+#include "rcutils/strcasecmp.h"
+
+int
+rcutils_strcasecmp(const char * s1, const char * s2)
+{
```
[the remark this answers is not in the document]
Comment (iuhilnehc-ynos) [reply src/strcasecmp.c:24]: And we can follow the style by
https://github.com/ros2/rcutils/blob/a287e74c700deb5816946349fa35aae31b1476f6/src/snprintf.c#L42-L47

File: include/rcutils/strcasecmp.h (hunk 2 above)
Comment (fujitatomoya) [reply include/rcutils/strcasecmp.h:23]: right, `.c` should be the right place.

File: src/strcasecmp.c (hunk 3 above)
Comment (fujitatomoya) [reply src/strcasecmp.c:18]: that is what i though too, but I'd like to keep it just in case.

File: include/rcutils/strcasecmp.h (hunk 4 above)
Comment (fujitatomoya) [reply include/rcutils/strcasecmp.h:44]: sgtm, will add `rcutils_strncasecmp` based on `strncasecmp` and `strnicmp`

File: test/test_strcasecmp.cpp (hunk 5 above)
Comment (fujitatomoya) [reply test/test_strcasecmp.cpp:21]: will do

File: src/strcasecmp.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (fujitatomoya) [reply src/strcasecmp.c:24]: yes, this is needed. will patch as you suggest.

**Comment by fujitatomoya** [comment]
@iuhilnehc-ynos @Ada-King 

comments are addressed, could you check when you got time?

File: src/strcasecmp.c [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  RCUTILS_CAN_FAIL_WITH({errno = EINVAL; return -1;});
+
+  if (s1 == NULL || s2 == NULL) {
+    errno = EINVAL;
+    return -1;
```
Comment (iuhilnehc-ynos) [inline src/strcasecmp.c:32]: I think we need to comment `errno` for this API, because the result '-1' might be a case that "s1" is less than "s2".

- The test case below expecting '-1' might be not good.

- About the test case, do you think we need to expect the return value in the cases is (-1, 0, 1), rather than 'NE 0'.

Comment (iuhilnehc-ynos) [reply src/strcasecmp.c:32]: Or maybe using LT GT is better.

File: include/rcutils/strcasecmp.h [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+/**
+ * This function compares two strings ignoring case which just wraps strcasecmp()
+ * or stricmp() in a portable way. This performs a byte-by-byte comparison of the
+ * strings s1 and s2, ignoring the case of the characters.
+ *
+ * \param[in] s1 String with null terminated to compare.
+ * \param[in] s2 String with null terminated to compare.
+ * \return An integer less than, equal to, or greater than zero if s1 is, after
+ *   ignoring case, found to be less than, to match, or be greater than s2,
+ *   respectively.
```
Comment (clalancette) [inline include/rcutils/strcasecmp.h:36]: I'll suggest we get rid of the discussion of the implementation.  It's a detail that the user of the API doesn't need to know.  How about:


```suggestion
 * This function compares two strings ignoring case in a portable way.
 * This performs a byte-by-byte comparison of the strings s1 and s2, ignoring the case of the characters.
 *
 * \param[in] s1 Null terminated string to compare.
 * \param[in] s2 Null terminated string to compare.
 * \return An integer less than, equal to, or greater than zero if s1 is, after
 *   ignoring case, found to be less than, to match, or be greater than s2,
 *   respectively.
```

File: include/rcutils/strcasecmp.h [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+ * or strnicmp() in a portable way. This performs a byte-by-byte comparison of the
+ * strings s1 and s2 up to count characters of s1 and s2, ignoring the case of the
+ * characters.
+ *
+ * \param[in] s1 String with null terminated to compare.
+ * \param[in] s2 String with null terminated to compare.
+ * \param[in] n Count characters to compare.
+ * \return An integer less than, equal to, or greater than zero if s1 is, after
+ *   ignoring case, found to be less than, to match, or be greater than s2,
+ *   respectively.
```
Comment (clalancette) [inline include/rcutils/strcasecmp.h:55]:
```suggestion
/// Case insensitive string compare up to count characters.
/**
 * This function compares two strings ignoring case in a portable way.
 * This performs a byte-by-byte comparison of the strings s1 and s2 up to count
 * characters of s1 and s2, ignoring the case of the
 * characters.
 *
 * \param[in] s1 First string to compare.
 * \param[in] s2 Second string to compare.
 * \param[in] n Count of characters to compare.
 * \return An integer less than, equal to, or greater than zero if s1 is, after
 *   ignoring case, found to be less than, to match, or be greater than s2,
 *   respectively.
```

It may also be worthwhile to talk about what happens if s1 is shorter than s2 and vice-versa.

File: src/strcasecmp.c (hunk 7 above)
Comment (clalancette) [reply src/strcasecmp.c:32]: Yeah, this is definitely a problem.  While there are some APIs that rely on errno for this (like `strtod`), it is hard to use them properly.  My suggestion here is to change it so that the return value is either 0 for success or -1 for failure.  Then we add in a pointer out argument which is the comparison value.  That way it is easy for the caller to tell whether it failed or not, and we disambiguate.

File: test/test_strcasecmp.cpp [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+  EXPECT_NE(0, rcutils_strcasecmp("abc1", "abc"));
+  EXPECT_NE(0, rcutils_strcasecmp("abc", "abc1"));
+  EXPECT_NE(0, rcutils_strcasecmp("ABCd", "abc"));
+  EXPECT_NE(0, rcutils_strcasecmp("abc", "ABCd"));
+  EXPECT_NE(0, rcutils_strcasecmp("1Abc", "abc"));
+  EXPECT_NE(0, rcutils_strcasecmp("abc", "1Abc"));
+  EXPECT_NE(0, rcutils_strcasecmp("a1Bc", "abc"));
+  EXPECT_NE(0, rcutils_strcasecmp("abc", "a1Bc"));
+  EXPECT_NE(0, rcutils_strcasecmp("ab1C", "abc"));
+  EXPECT_NE(0, rcutils_strcasecmp("abc", "ab1C"));
```
Comment (clalancette) [inline test/test_strcasecmp.cpp:61]: These tests should be more specific; instead of just `NE`, they should check an exact value:

```
EXPECT_EQ(-1, rcutils_strcasecmp("a", "b");
```

Or similar.

File: src/strcasecmp.c (hunk 7 above)
Comment (fujitatomoya) [reply src/strcasecmp.c:32]: @clalancette 

that makes sense to me.

**Comment by fujitatomoya** [comment]
@clalancette @iuhilnehc-ynos @Ada-King 

could you do review again?

**Review by iuhilnehc-ynos — APPROVED** [review]

**Review by Ada-King — APPROVED** [review]

File: include/rcutils/strcasecmp.h [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+/// Case insensitive string compare.
+/**
+ * This function compares two strings ignoring case in a portable way.
+ * This performs a byte-by-byte comparison of the strings s1 and s2,
+ * ignoring the case of the characters.
+ *
+ * \param[in] s1 Null terminated string to compare.
+ * \param[in] s2 Null terminated string to compare.
+ * \param[out] value Pointer to camparison result.
```
Comment (clalancette) [inline include/rcutils/strcasecmp.h:34]:
```suggestion
 * \param[out] value Pointer to comparison result.
```

File: include/rcutils/strcasecmp.h [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+ *   An integer less than, equal to, or greater than zero if s1 is, after
+ *   ignoring case, found to be less than, to match, or be greater than s2,
+ *   respectively.
+ * \return 0 if s1 and s2 match, -1 if don't.
```
Comment (clalancette) [inline include/rcutils/strcasecmp.h:38]: I don't think this is what we want.  The result of the comparison is already stored in `*value`; this should just be a return of 0 if the method succeeded and -1 if it failed.

File: include/rcutils/strcasecmp.h [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+/// Case insensitive string compare up to count characters.
+/**
+ * This function compares two strings ignoring case in a portable way.
+ * This performs a byte-by-byte comparison of the strings s1 and s2 up to count
+ * characters of s1 and s2, ignoring the case of the characters.
+ *
+ * \param[in] s1 First string to compare.
+ * \param[in] s2 Second string to compare.
+ * \param[in] n Count of characters to compare.
+ * \param[out] value Pointer to camparison result.
```
Comment (clalancette) [inline include/rcutils/strcasecmp.h:57]:
```suggestion
 * \param[out] value Pointer to comparison result.
```

File: include/rcutils/strcasecmp.h [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+ *   An integer less than, equal to, or greater than zero if s1 is, after
+ *   ignoring case, found to be less than, to match, or be greater than s2,
+ *   respectively.
+ * \return 0 if s1 and s2 match, -1 if don't.
```
Comment (clalancette) [inline include/rcutils/strcasecmp.h:61]: I don't think this is what we want. The result of the comparison is already stored in *value; this should just be a return of 0 if the method succeeded and -1 if it failed.

File: src/strcasecmp.c [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+#ifndef _WIN32
+  *value = strcasecmp(s1, s2);
+#else
+  *value = stricmp(s1, s2);
+#endif
+  if (*value == 0) {
+    return 0;
+  } else {
+    return -1;
+  }
```
Comment (clalancette) [inline src/strcasecmp.c:43]:
```suggestion
  return 0;
```

File: src/strcasecmp.c [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+#ifndef _WIN32
+  *value = strncasecmp(s1, s2, n);
+#else
+  *value = strnicmp(s1, s2, n);
+#endif
+  if (*value == 0) {
+    return 0;
+  } else {
+    return -1;
+  }
```
Comment (clalancette) [inline src/strcasecmp.c:65]:
```suggestion
  return 0;
```

File: include/rcutils/strcasecmp.h (hunk 11 above)
Comment (fujitatomoya) [reply include/rcutils/strcasecmp.h:34]: my bad. thanks

File: include/rcutils/strcasecmp.h (hunk 12 above)
Comment (fujitatomoya) [reply include/rcutils/strcasecmp.h:38]: actually i was wondering that. At the moment i was thinking that maybe it is simpler to return 0 if strings are the same ignoring cases. (it is likely to use this way, i guess) but i understand that this changes behavior of `strcasecmp` or `stricmp`. After all, taking your suggestion.

**Comment by fujitatomoya** [comment]
@clalancette 

thanks for the suggestion, i changed it back along with `strcasecmp` behavior. could you check when you got time?

**Review by clalancette — APPROVED** [review]

**Comment by clalancette** [comment]
CI (only up to rcutils since this is a new API):
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12206/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7189/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9932/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12095/)

**Comment by clalancette** [comment]
@fujitatomoya It looks like Windows has some warnings with this change.  Can you look into it?  Thanks.

**Comment by fujitatomoya** [comment]
@clalancette 

got it, i will address them in today. will get back to you.

**Comment by fujitatomoya** [comment]
@clalancette 

i believe windows is now comfortable with https://github.com/ros2/rcutils/pull/280/commits/9b586f246c59a111ac238ec664babec9506fb7ff. Could you run CI again? thank!

**Comment by clalancette** [comment]
Thanks.  Here's another CI run:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12220/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7203/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9946/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12115/)

**Comment by clalancette** [comment]
Looks good now, thanks for iterating @fujitatomoya !
